src/python/zquantum/core/interfaces/ansatz.py

```diff
@@ -0,0 +1,79 @@
+from abc import ABC, abstractmethod
+from ..circuit import Circuit
+from .ansatz_utils import ansatz_property
+from ..utils import create_symbols_map
+import copy
+import sympy
+from typing import List, Optional
+from overrides import EnforceOverrides
+import numpy as np
+
+
+class Ansatz(ABC, EnforceOverrides):
+
+    supports_parametrized_circuits = None
+    n_layers = ansatz_property("n_layers")
+
+    def __init__(self, n_layers: int):
+        """
+        Interface for implementing different ansatzes.
+        This class also caches the circuit for given ansatz parameters.
+
+        Args:
+            n_layers: number of layers of the ansatz
+        
+        Attributes:
+            n_layers (int): see Args
+            circuit (zquantum.core.circuit.Circuit): circuit representation of the ansatz.
+            supports_parametrized_circuits(bool): flag indicating whether given ansatz supports parametrized circuits.
+        
+        """
+        self.n_layers = n_layers
+        self._parametrized_circuit = None
+
+    @property
+    def parametrized_circuit(self) -> Circuit:
+        if self._parametrized_circuit is None:
+            if self.supports_parametrized_circuits:
+                return self._generate_circuit()
+            else:
+                raise (
+                    Exception(
+                        "{0} does not support parametrized circuits.".format(
+                            type(self).__name__
+                        )
+                    )
+                )
+        else:
+            return self._parametrized_circuit
+
+    def get_executable_circuit(self, parameters: np.ndarray) -> Circuit:
+        """
+        Returns an executable circuit representing the ansatz.
+        Args:
+            parameters: circuit parameters
+        """
+        if parameters is None:
+            raise (Exception("Parameters can't be None for executable circuit."))
```

Comment:
How come `parameters` is a required argument? Shouldn't this be optional in the case of an ansatz without parameterized gates?

Reply:
I think I don't get what you mean.
1. This is "executable" circuit, so it cannot have any parameters in the circuit.
2. If there would be an ansatz without any parameters (which I think kills the purpose) I think you'll just pass an empty array?

Reply:
1. Correct, I meant defining a circuit without parameterizable gates.
2. Fair point. I was thinking more of an Ansatz as a circuit template/circuit factory, but not under the overarching context of being used in variational algorithms.